keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+
+When an Application object is deleted, the Garbage Collector will detect that the owner of the Application's components 
+no longer exists at the API Server. This will trigger deletion of the children objects. Note that the Application 
+Controller does not establish an ownership relationship between an Application and its dependencies. This is by design.
+Also note that, if the Application is composed of Kubernetes primitives that, in turn, aggregate other objects, it is 
+sufficient to only add an OwnerReference to the top level object. For instance, deleting a Application containing a 
+Deployment will delete the Deployment's ReplicaSets and Pods.
+
+#### Installation Status Reporting
+The Application Controller will additionally update the `Status` of Application objects. When the Application Controller 
+detects newly created objects, for all objects other than Applications, it shall do the following.
+
+1. [Retrieve the Application for the object](#retrieving-applications).
+1. If the Application's `Selector` matches the object, the Application Controller shall update the Applications status 
+to indicate that component is installed.
+
+For all Application objects, the Application Controller shall do the following upon creation.
+
+1. [Select the Application's components](#selecting-components).
+1. For each selected object, the Application Controller will update the Application object's `Status` to indicate that 
+the component is installed. If a component is not found, the Applications status will be updated to indicate that the 
+component is not installed.
+1. [Retrieve the Application's dependencies](#retrieving-dependencies).
+1. For each dependency, the Application controller will update the the Application object's `Status` to indicate the 
+installation status of the dependency.
+1. The Application Controller will then scan all installed Applications in the same namespace as the Application 
+under consideration. If any of those Applications list the Application under consideration in their `Spec.Dependencies`, 
+their `Status` will be updated to indicate that the dependency has been installed.
+
+For all objects other than Applications, upon object deletion, the Application Controller shall do the following. 
+
+1. [Retrieve the Application for the object](#retrieving-applications).
+1. Update the Application's `Status` to indicate that component is no longer installed.
+
+For all Application objects, upon deletion, the Application Controller will do the following.
+
+1. List all Applications in the same namespace as the application under consideration.
+1. For all Applications that contain a reference to the Application under consideration in their `Depepndecies`, the 
+Application Controller will update the Application's `Statuts` to indicate that the Application under consideration is 
+no longer installed.
```

Comment:
And if a anything fails to delete, what happens? The global state of the application does not seem to have a capacity to indicate that we are in a failed state.

In the case of a dependency, it does not seem that the application object has any capability to bubble up the fact that a dependency was deleted, either.

Both of these seem to require that a user first happen to notice the failure (since there's no top-level indication of an error on an Application object), and then do quite a bit of manual debugging to find out what happened.

Reply:
If we want to leverage Kubernetes GC, as all of the core controllers do, this is a fundemental limitation. However, if we leverage Kubernetes GC, deletion failure is a global bug in the system.